Automattic-Tracks-iOS/CrashLogging.swift

```diff
@@ -0,0 +1,156 @@
+        /// If we can't say for sure, assume the user has opted out
+        guard let dataProvider = sharedInstance.dataProvider else { return true }
+        return dataProvider.userHasOptedOut
+    }
+
+    /// Immediately crashes the application and generates a crash report.
+    public static func crash() {
+        Client.shared?.crash()
+    }
+}
+
+// Manual Error Logging
+public extension CrashLogging {
+
+    /**
+     Writes the error to the Crash Logging system, and includes a stack trace.
+
+     - Parameters:
+     - error: The error object
+     - userInfo: A dictionary containing additional data about this error.
+     - level: The level of severity to report in Sentry (`.error` by default)
+    */
+    static func logError(_ error: Error, userInfo: [String : Any]? = nil, level: SentrySeverity = .error) {
+        let event = Event(level: .error)
+        event.message = error.localizedDescription
+        event.extra = userInfo ?? (error as NSError).userInfo
+
+        Client.shared?.snapshotStacktrace {
+            Client.shared?.appendStacktrace(to: event)
+        }
+
+        Client.shared?.send(event: event)
+        sharedInstance.dataProvider?.didLogErrorCallback?(event)
+    }
+
+    /**
+     Writes a message to the Crash Logging system, and includes a stack trace.
+     - Parameters:
+     - message: The message
+     - properties: A dictionary containing additional information about this error
+     - level: The level of severity to report in Sentry (`.error` by default)
+    */
+    static func logMessage(_ message: String, properties: [String : Any]? = nil, level: SentrySeverity = .info) {
+
+        let event = Event(level: level)
+        event.message = message
+        event.extra = properties
+
+        Client.shared?.snapshotStacktrace {
+            Client.shared?.appendStacktrace(to: event)
+        }
+
+        Client.shared?.send(event: event)
+        sharedInstance.dataProvider?.didLogMessageCallback?(event)
+    }
+}
+
+// User Tracking
+extension CrashLogging {
+
+    internal var currentUser: Sentry.User {
+
+        let anonymousUser = TracksUser(userID: nil, email: nil, username: nil).sentryUser
+
+        /// Don't continue unless `start` has been called on the crash logger
+        guard let dataProvider = self.dataProvider else { return anonymousUser }
+
+        /// Don't continue if the data source doesn't yet have a user
+        guard let user = dataProvider.currentUser else { return anonymousUser }
+        let data = dataProvider.additionalUserData
+
+        return user.sentryUser(withData: data)
+    }
+
+    /// Causes the Crash Logging System to refresh its knowledge about the current user.
+    ///
+    /// This is required in situations like login / logout, when the system otherwise might not
+    /// know a change has occured.
+    ///
+    /// Calling this method in these situations prevents
+    public static func updateCurrentUser() {
```

Comment:
Your point is 💯, but I'm not sure about the name 🤔 

The data provider hasn't necessarily _changed_, because ideally it'd just call out to other classes without storing much (if any) internal state.

`setDataProviderHasNewDataAvailable` covers that, but is...verbose. We could do `setNewDataAvailable` but then it might be less clear to the reader what'll happen when they call it? (the inline docs could help with that...?)

Also, the `set...` syntax seems _slightly_ more at home on iOS (I'm thinking of things like `setNeedsLayout`, as a for-instance), whereas `notify` might not be as common?

There's a really good chance I'm overthinking this naming thing 🤔😅

Reply:
Follow-up to this, `setNeedsDataRefresh` seems like a nice framework-integrated name.

Sorry, I'll stop 🤓-ing out about names now.

Reply:
Yeah. You know what? Let's leave it as is for now until we find a better name. 😄

Reply:
@jkmassel Oh, I did not see the latest `setNeedsDataRefresh`! Let's do that. That's a great name. 👍